vendor/bat-native-ledger/src/bat/ledger/internal/gemini/gemini_util.cc

```diff
@@ -0,0 +1,269 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include <utility>
+
+#include "base/base64.h"
+#include "base/json/json_reader.h"
+#include "base/json/json_writer.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/strings/string_util.h"
+#include "base/strings/stringprintf.h"
+#include "bat/ledger/global_constants.h"
+#include "bat/ledger/internal/common/random_util.h"
+#include "bat/ledger/internal/gemini/gemini_util.h"
+#include "bat/ledger/internal/ledger_impl.h"
+#include "bat/ledger/internal/state/state_keys.h"
+#include "crypto/random.h"
+
+namespace ledger {
+namespace gemini {
+
+std::string GetClientId() {
+  return ledger::_environment == type::Environment::PRODUCTION
+             ? GEMINI_WALLET_CLIENT_ID
+             : GEMINI_WALLET_STAGING_CLIENT_ID;
+}
+
+std::string GetClientSecret() {
+  return ledger::_environment == type::Environment::PRODUCTION
+             ? GEMINI_WALLET_CLIENT_SECRET
+             : GEMINI_WALLET_STAGING_CLIENT_SECRET;
+}
+
+std::string GetUrl() {
+  return ledger::_environment == type::Environment::PRODUCTION
+             ? GEMINI_OAUTH_URL
+             : GEMINI_OAUTH_STAGING_URL;
+}
+
+std::string GetFeeAddress() {
+  return ledger::_environment == type::Environment::PRODUCTION
+             ? kFeeAddressProduction
+             : kFeeAddressStaging;
+}
+
+std::string GetACAddress() {
+  return ledger::_environment == type::Environment::PRODUCTION
+             ? kACAddressProduction
+             : kACAddressStaging;
+}
+
+std::string GetAuthorizeUrl(const std::string& state) {
```

Comment:
This is also copied. Using the same method for all of them is trivial.

Reply:
Every custodian has an authorize URL, but the parameters to construct the URL are different. 

Uphold - `state` and `kyc_flow`
Gemini - `state`
Bitflyer - `state` and `code_verifier`

Also, each `GetAuthorizeUrl` specifies a different set of scopes required for each custodian.

Reply:
I was able to generalize `Get/Set/Reset wallet` for different custodians - https://github.com/brave/brave-core/pull/8566/files#diff-5ec1ba12eb29291f5c93c724e53cf6c4f6acfce0d56681866d54fef6e990ca5a

Will update the code to use JSONValueConverter